# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFocusManager.cpp
# Commit: 92148b810f22
# Full Hash: 92148b810f228e4e9e84255682fe888a5bee07d5
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-12-24 09:09:23
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Bug 779684 - part 2: Get rid of `nsFrameTraversal` and `nsIFrameTraversal` r=emilio
#   
#   The class is used only for creating `nsFrameIterator`, but it's unnecessary
#   so that we can get rid of it and its interface.
#   
# ==============================================================================

diff -r d4411b38cae7 -r 92148b810f22 dom/base/nsFocusManager.cpp
--- a/dom/base/nsFocusManager.cpp	Sun Dec 24 06:01:19 2023 +0000
+++ b/dom/base/nsFocusManager.cpp	Sun Dec 24 06:01:19 2023 +0000
@@ -3220,19 +3220,20 @@
       text && text->TextDataLength() == domRange->StartOffset() &&
       domSelection->IsCollapsed()) {
     nsIFrame* startFrame = start->GetPrimaryFrame();
+    MOZ_ASSERT(startFrame);
     // Yes, indeed we were at the end of the last node
-    RefPtr<nsFrameIterator> frameIterator;
     nsIFrame* limiter =
         domSelection && domSelection->GetAncestorLimiter()
             ? domSelection->GetAncestorLimiter()->GetPrimaryFrame()
             : nullptr;
-    MOZ_TRY(NS_NewFrameTraversal(getter_AddRefs(frameIterator), presContext,
-                                 startFrame, eLeaf,
-                                 false,  // aVisual
-                                 false,  // aLockInScrollView
-                                 true,   // aFollowOOFs
-                                 false,  // aSkipPopupChecks
-                                 limiter));
+    RefPtr<nsFrameIterator> frameIterator = nsFrameIterator::Create(
+        presContext, startFrame, nsFrameIterator::Type::Leaf,
+        false,  // aVisual
+        false,  // aLockInScrollView
+        true,   // aFollowOOFs
+        false,  // aSkipPopupChecks
+        limiter);
+    MOZ_ASSERT(frameIterator);
 
     nsIFrame* newCaretFrame = nullptr;
     nsIContent* newCaretContent = start;
@@ -4198,14 +4199,14 @@
       // For tab navigation, pass false for aSkipPopupChecks so that we don't
       // iterate into or out of a popup. For document naviation pass true to
       // ignore these boundaries.
-      nsresult rv = NS_NewFrameTraversal(
-          getter_AddRefs(frameIterator), presContext, frame, ePreOrder,
+      frameIterator = nsFrameIterator::Create(
+          presContext, frame, nsFrameIterator::Type::PreOrder,
           false,                  // aVisual
           false,                  // aLockInScrollView
           true,                   // aFollowOOFs
           aForDocumentNavigation  // aSkipPopupChecks
       );
-      NS_ENSURE_SUCCESS(rv, rv);
+      MOZ_ASSERT(frameIterator);
 
       if (iterStartContent == aRootContent) {
         if (!aForward) {
@@ -4414,7 +4415,7 @@
             currentContent->IsHTMLElement(nsGkAtoms::img) &&
             currentContent->AsElement()->HasAttr(nsGkAtoms::usemap)) {
           // This is an image with a map. Image map areas are not traversed by
-          // nsIFrameTraversal so look for the next or previous area element.
+          // nsFrameIterator so look for the next or previous area element.
           nsIContent* areaContent = GetNextTabbableMapArea(
               aForward, aCurrentTabIndex, currentContent->AsElement(),
               iterStartContent);